quickcheck_derive/src/structural.rs

```diff
@@ -0,0 +1,170 @@
+use quote::Tokens;
+use syn::{DeriveInput, Field, Ident, Variant, VariantData};
+
+pub fn derive_struct(item: &DeriveInput, variant: &VariantData) -> (Tokens, Tokens) {
+    let name = &item.ident;
+    
+    let arbitrary = arbitrary_variant(variant, name);
+    let shrink = match *variant {
+        VariantData::Struct(ref fields) => {
+            let field_names = fields.iter()
+                .map(|f| f.ident.as_ref().unwrap())
+                .collect::<Vec<_>>();
+            let field_names = &field_names;
+            let alphas = alpha_names(fields.len());
+            let alphas = &alphas;
+
+            let tuple_pattern = match alphas.len() {
+                0 => quote!(()),
+                1 => {
+                    let alpha = &alphas[0];
+                    quote!(#alpha)
+                },
+                _ => quote!((#(#alphas),*)),
+            };
+
+            quote! {
+                Box::new(
+                    (#(self.#field_names),*).shrink().map(|#tuple_pattern| #name {
+                        #(#field_names: #alphas),*
+                    })
+                )
+            }
+        },
+        VariantData::Tuple(ref fields) => {
+            let field_names = (0..fields.len()).map(Ident::new).map(|i| quote!(self.#i));
+            let alpha_names = &alpha_names(fields.len());
+
+            quote! {
+                // TODO This isn't a *great* way to do this until we get
+                // generics over tuples, to be able to implement shrinking
+                // for tuples of all sizes.
+                Box::new((#(#field_names),*).shrink().map(|(#(#alpha_names),*)| #name(#(#alpha_names),*)))
```

Comment:
This looks really useful! Attempted to use it in a project of mine and discovered a bug with newtypes. `(#(#alpha_names),*)` needs to get the same treatment as `#tuple_pattern` above or it fails for `struct T(i32);`.

(Accidently deleted this comment).